src/channeldef.ts

```diff
@@ -198,6 +199,87 @@ export type TextDef<F extends Field> =
   | FieldOrDatumDefWithCondition<StringDatumDef<F>, Text>
   | ValueDefWithCondition<StringFieldDef<F>, Text>;
 
+/**
+ * Position of a label relative to its base mark's position, which includes anchor direction (anchor) and distance (offset).
+ */
+export type LabelPosition = {
+  /**
+   * Distance of a label from its base mark.
+   * If negative, the label position is insize.
+   */
+  offset: number;
+
+  /**
+   * Anchor direction of a label from its base mark.
+   * The value could be one of `"left"`, `"right"`, `"top"`, `"middle",`
+   * `"bottom"`, `"top-left"`, `"top-right"`, `"bottom-left"`, or "`bottom-right"`.
+   */
+  anchor: LabelAnchor;
+};
+
+type LabelMarkDef = Omit<MarkDef<'text'>, 'type'>;
+
+export type LabelDefMixins = {
+  /**
+   * A list of possible positions for a label to be placed relative to its base mark.
+   */
+  position?: LabelPosition[];
+
+  /**
+   * The level of ancestor layers of marks that the labels will avoid.
+   * `'all'` means all the marks that are children of the root layer.
+   * Otherwise, a number `n` means all the marks that are children of the nth ancestor layer.
+   *
+   * __Default value:__ `0` (only avoid the base mark)
+   */
+  avoidAncestorLayer?: 'all' | number;
+
+  /**
+   * A label mark definition for customizing the label's text mark.
+   * This definition includes all [properties of a text mark](http://vega.github.io/vega-lite/docs/text.html#properties) except `type`.
+   */
+  mark?: LabelMarkDef;
+
+  /**
+   * The padding in pixels by which a label may extend past the chart bounding box.
+   */
+  padding?: number;
+
+  /**
+   * The labeling method to use for area marks.
+   * One of `"floodfill"`, `"reduced-search"`, and `"naive"`.
```

Comment:
For both one of here and the explanation paragraphs below, I'd begin with "reduced-search" since it's the default.